### DISCUSSION TITLE
Casadi error with allow_free, while using do-mpc

### DISCUSSION THREAD

**Discussion by YuganshuWadhwa** [do-mpc/do-mpc#416] [Q&A]
Hello all, 

I am trying to implement a MPC for a physical system. for which I have a continuous state-space model (represented as discrete here) of the form,

`x(k+1) = A . x(k) + B . u(k)`
`y(k) = C . x(k) + D . u(k)`

where the dimensions of the matrices are: `A = 4x4`, `B = 4x1`, `C = 2x4`, and `D = 2x1`.

This is a single input (1) multi output (2) model, and was derived using another system identification package. Therefore any knowledge about the states `x_0`, `x_1`, `x_2`, `x_3` is unavailable. And therefore no information about the actual model physics is available. 

So far I have modeled this by declaring `u` as `var_type='_u'`, and `x_0`, `x_1`, `x_2`, `x_3` as `var_type='_x'`. I also tried declaring a single state `x` with a `4x1` shape but the error persists. 

My first question would be, ***how do I declare the output equation for `y(k)` ?*** 
Should I use 'set_meas` or `set_expression`?

Coming to the main issue, I want to use one of the outputs (`y_0` for example) to implement set-point tracking. For this I use another time-varying parameter  `y_0_set` as my reference. I do this so that I can use, 

`mterm = ( model.tvp['y_0_set'] - y_0 ) ** 2` where `y_0` comes from the output expression set above (which again, I am unsure how to declare). 

When I currently try to do this, I get an error saying:
`mterm::init: Initialization failed since variables [x_0x_1x_2x_3] are free. These symbols occur in the output expressions but you forgot to declare these as inputs. Set option 'allow_free' to allow free variables.`

And I am looking for any fixes to this issue. I have tried looking through the documentation (both do-mpc and casadi) on how to explicitly trigger the `allow_free` parameter, but no luck so far.

Any insights regarding this issue, or just regarding my approach or my model is appreciated. I am quite new to this library so any information is welcome. 

Thanks :)

**Comment by miroder** [comment]
```
x = model.set_variable(var_type='_x', var_name='x', shape=(proc_ss.A.shape[0],1))
u = model.set_variable(var_type='_u', var_name='u', shape=(1,1))
x_next = proc_ss.A@x + proc_ss.B@u
model.set_rhs('x', x_next)
y = model.set_expression('y', proc_ss.C@x + proc_ss.D@u)
cost = model.set_expression('cost', (y-model.tvp['set_point'])**2)
...
mpc.set_objective(mterm=model.aux['cost'], lterm=model.aux['cost'])
```


try `model.aux['y_0']` instead of 'y'

**Comment by miroder** [comment]
```
x = model.set_variable(var_type='_x', var_name='x', shape=(proc_ss.A.shape[0],1))
u = model.set_variable(var_type='_u', var_name='u', shape=(1,1))
x_next = proc_ss.A@x + proc_ss.B@u
model.set_rhs('x', x_next)
y = model.set_expression('y', proc_ss.C@x + proc_ss.D@u)
cost = model.set_expression('cost', (y-model.tvp['set_point'])**2)
...
mpc.set_objective(mterm=model.aux['cost'], lterm=model.aux['cost'])
```


try `model.aux['y_0']` instead of 'y'

**Comment by YuganshuWadhwa** [comment]
@miroder Thanks for the suggestions. I'll keep them in mind!!
